### ISSUE TITLE
Use packages instead of ports ?

### ISSUE THREAD

**Issue by bozhinov** [iredmail/iRedMail#272]
REQUIRED BASIC INFO OF YOUR IREDMAIL SERVER:

- iRedMail version (check /etc/iredmail-release):  1.7.1
- Deployed with iRedMail Easy or the downloadable installer? No
- Linux/BSD distribution name and version: Freebsd 14.2
- Store mail accounts in which backend (LDAP/MySQL/PGSQL):  LDAP
- Web server (Apache or Nginx): Nginx
- Manage mail accounts with iRedAdmin-Pro? No
- [IMPORTANT] Related original log or error message is required if you're experiencing an issue.

Hello,

I have not used ports in ages. I did not used to mind it but
- 3 Gb of ports take a while to compile
- I chose LDAP for a backend and it ended up downloading MySQL & PGSQL any way along with all of their dependencies 

If all is going to be installed anyway - wouldn't it be much better to use packages ?

**Comment by iredmail** [comment]
We tried years ago, but pre-built packages don't have required compile options. e.g. LDAP support, MySQL support, etc.
Not sure how it works nowadays.

**Comment by bozhinov** [comment]
the same say
it should be possible to install cmake, mariadb and rust at least from pkg.
that will save like couple of hours and my disks.

**Comment by iredmail** [comment]
Mixing ports and packages is not good idea. 
if you don't want to wait for compile, try Linux or OpenBSD instead. :(

**Comment by bozhinov** [comment]
I have considered OpenBSD but there is a lot more running on that machine and who has the time to patch Linux boxes as often as needed.

I m sure most admins would deploy iRedMail in a jail that will be used for that purpose alone.
Long term maintenance using ports is no picnic either

**Comment by iredmail** [comment]
Hi @bozhinov 

Would you like to give it a try? Update `functions/packages_freebsd.sh` to install packages with `pkg` instead of ports tree.

**Comment by bozhinov** [comment]
I have added you to my private repo.
apart from that stupid SSL bug it is OK
says fixed in FreeBSD 15-dev branch

**Comment by iredmail** [comment]
Hi @bozhinov 

could you fork the iRedMail repo directly instead of copying repo and make it private? So that we can easily create pull request after finished.

**Comment by bozhinov** [comment]
Hi,
I m definitely doing that. Need few days. My schedule is hectic these days.

**Comment by xbeaudouin** [comment]
Hello there,
On ports there is now flavors.
As a FreeBSD ports commiter I can adapt (if needed) the ports to have such flavors already done.
Also I can help to have a mail/iredmail port as well.
What do you think of that?
Kind regards

**Comment by iredmail** [comment]
Hi @xbeaudouin 

That's wonderful. 

Flavors should be enough. We need few packages with MySQL, PostgreSQL, OpenLDAP support. For example, postfix, Dovecot, Amavisd.  If you can help make it happen, we can work together. :)

**Comment by xbeaudouin** [comment]
Well dovecot has already flavors. I am trying to understand `functions/packages_freebsd.sh` and have a check about current port tree.
Really I would like this project works to replace my old mail server.

**Comment by xbeaudouin** [comment]
By the way why should you need openssl from ports? Since on FreeBSD 14.2 we already have openssl v3.0.15:
```
$ openssl version
OpenSSL 3.0.15 3 Sep 2024 (Library: OpenSSL 3.0.15 3 Sep 2024)
```

**Comment by iredmail** [comment]
Do you mean we should remove `ssl=openssl` from `/etc/make.conf`?
https://github.com/iredmail/iRedMail/blob/master/functions/packages_freebsd.sh#L54

**Comment by iredmail** [comment]
According to functions/packages_freebsd.sh, these ports requires flavors:

- mail/spamassassin: need flavors for mysql and postgresql support. https://github.com/iredmail/iRedMail/blob/master/functions/packages_freebsd.sh#L476-L480
- mail/roundcube: need flavors for openldap, mysql, postgresql. Note: for iRedMail OpenLDAP backend, it requires both openldap and mysql support.
- mail/postfix: need flavors for openldap, mysql, postgresql. Note: for iRedMail OpenLDAP backend, it requires both openldap and mysql support.
- security/amavisd-new: need flavors for openldap, mysql, postgresql. Note: for iRedMail OpenLDAP backend, it requires both openldap and mysql support.
- devel/apr1: need flavors for openldap, mysql, postgresql. Note: for iRedMail OpenLDAP backend, it requires both openldap and mysql support.
- devel/sope: need flavors for openldap, mysql, postgresql. Note: for iRedMail OpenLDAP backend, it requires both openldap and mysql support.
- databases/py-sqlalchemy14: need flavors for mysql, postgresql.

**Comment by iredmail** [comment]
Currently the installation involves compiling Rust / Go / LLVM, you can imagine how much time we waste.

**Comment by xbeaudouin** [comment]
Hi,
Well excepted if you need some special cyphers on openssl, I don't see any good reason to add this dependency. Postfix, nginx, dovecot can be run on freebsd stock openssl AFAIK. So I think we can remove ssl=openssl in make.conf.

About ports with flavors, I will try to push some of them. Issue will be "dual" flavors which is not supported by the port infrastructure

**Comment by iredmail** [comment]
If dual flavors is impossible, then we still have to install all packages from ports tree, to avoid mixing binary packages and ports.
Is such mixing recommended or not?

I see Dovecot has LDAP flavor on by default, and you can choose additional MySQL or PGSQL flavor. Is it possible that we have LDAP flavor on by default too for other ports?
For example, MySQL flavor of `mail/postfix` supports both LDAP and MySQL, PGSQL flavor of `mail/postfix` supports both LDAP and PGSQL.

**Comment by iredmail** [comment]
Or, besides standard `ldap`, `mysql`, `pgsql` flavers, can we create `ldap-mysql`, `ldap-pgsql` flavor?
I mean, `ldap-mysql` is just one flavor name, except it enables two compile options instead of one.

- in `FLAVORS=`, append `ldap-mysql` as a flavor name:
https://cgit.freebsd.org/ports/tree/mail/postfix/Makefile?id=d4e949bafdb4e842868267d26a357c9921d3efeb#n25

- Add `ldap_mysql_PKGNAMESUFFIX=		-ldap-mysql`

**Comment by xbeaudouin** [comment]
the '-' is used to sperate the flavors... I will discuss with other maintainer for that...

**Comment by xbeaudouin** [comment]
I think we can mix both packages and compiled with port ones. There is no problems about that.
Also the LDAP can be the exception.
Maybe we should do a iredmail ports, that will handle that too.

**Comment by xbeaudouin** [comment]
I think we can make some "special flavors" eg -mysqlldap -pgsqlldap for such ports.

**Comment by iredmail** [comment]
I will wait for your update. Thank you very much. 

Special flavor might be a trick or temporary solution, for example,
What if someone needs ldap + mysql + pgsql? I wish FreeBSD ports tree supports multi flavors like OpenBSD in the future. :)

**Comment by bozhinov** [comment]
Not sure what happened here but I m not compiling this from ports ever again.

https://github.com/bozhinov/iRedMail-pub/commit/66bb775fc3861396a3d24d78dc65912e80aff102

Turned out MariaDb is not possible as a backend on FreeBSD
There is no postfix-mariadb package

**Comment by iredmail** [comment]
> Got it working 100% with PGSQL via packages https://github.com/bozhinov/iRedMail-pub/commits/master/

Great news. 👍 
Could you please use only one `pkg install` command to install all packages?

**Comment by bozhinov** [comment]
Done!
@iredmail I initially cloned the masted branch. at some point \n became \r\n
could have been me though. I run on Windows.

@xbeaudouin adding support back for MariaDb does not seem like a lot of work
these 3 ports could easily 3 new ones for MariaDb:
dovecot-MySQL
dovecot-pigeonhole-MySQL
postfix-mysql

**Comment by xbeaudouin** [comment]
@bozhinov Nice :) Well did you made a PR ?

@iredmail if @bozhinov made a PR, can you merge it ? Then I will update for mysql/mariadb pkg ?

**Comment by iredmail** [comment]
How about I merge it in a new git branch then we continue in that branch instead of master?

**Comment by bozhinov** [comment]
Can't push with those new lines all messed up. thanks to the windows desktop client
@iredmail whatever works. the dialog on FreeBSD will still show MariaDb & ldap as supported

**Comment by iredmail** [comment]
Hi @bozhinov 

Could you please grant me write privilege in your repo? I made some changes locally and waiting for push to your repo.

**Comment by bozhinov** [comment]
sent invite

**Comment by iredmail** [comment]
Hi @bozhinov 

Not necessary.
I created branch `pkg` with all your changes, and my improvements. Thanks.
https://github.com/iredmail/iRedMail/pull/280

Your changes dropped support for other backends, i have to clean it up.

**Comment by bozhinov** [comment]
I used latest for dev test and I use latest in production

**Comment by iredmail** [comment]
Only Latest packages offer packages like `dovecot-mysql`, `dovecot-pgsql`.
We have to use Latest for now.

**Comment by iredmail** [comment]
Improved with MariaDB backend support in PR #280 
Currently installation goes well for both mariadb and pgsql backends, but didn't have time to test applications yet.
Please help test.

**Comment by bozhinov** [comment]
Hi,

 

postfix-mysql this package forces the removal of MariaDb

tested already

 

won’t work

 

Momchil

**Comment by xbeaudouin** [comment]
Well this is mostly because -mysql packages on the ports `USE=mysql` in `Mk/bsd.default-version.mk` use : 
```
# Possible values: 8.0, 8.4, 9.0, 9.1, 10.5m, 10.6m, 10.11m, 11.4m
MYSQL_DEFAULT?=		8.0
```

If the poudriere that make this set mysql:10.6m then all packages are set to mariadb...
This is also one of the reasons I want to make a port of the installer.

**Comment by bozhinov** [comment]
I imagine a port would break these guises' business model

I wonder what happens when MySQL reaches v10 in another 5 to 10 years.

**Comment by iredmail** [comment]
> postfix-mysql this package forces the removal of MariaDb

Thanks for helping. I found this conflict moment ago during testing, yesterday it's kind of rush and didn't notice the conflict message reported by `pkg`.

**Comment by iredmail** [comment]
> Well this is mostly because -mysql packages on the ports `USE=mysql` in `Mk/bsd.default-version.mk` use :
> 
> ```
> # Possible values: 8.0, 8.4, 9.0, 9.1, 10.5m, 10.6m, 10.11m, 11.4m
> MYSQL_DEFAULT?=		8.0
> ```

Hi @xbeaudouin 

All packages with `-mysql` flavor require MySQL client package (e.g. `mysql80-client`), but iRedMail uses MariaDB instead for a long time.

Possible solutions:

1. Switch to MySQL on FreeBSD (no plan for other linux/openbsd).
2. Tune FreeBSD `pkg` to use MariaDB instead of MySQL. I didn't find any doc, guess it's impossible. Ideally packages with `-mysql` flavor should work with either mariadb or mysql, but mysql is now forced, and many people don't think mariadb is a drop-in replacement for mysql in some cases.
3. Improve FreeBSD ports tree to add `-mariadb` flavor for all these packages. for example, `postfix-mariadb`, `dovecot-mariadb`.

```
# pkg install ...
...
Checking integrity... done (1 conflicting)
  - mysql80-client-8.0.39 conflicts with mariadb106-client-10.6.20 on /usr/local/bin/my_print_defaults
```

**Comment by iredmail** [comment]
Switched to MySQL in PR moment ago.
- [x] iRedMail installation for MySQL backend went fine
- [x] Roundcube works fine

@xbeaudouin Need your help:

- SOGo doesn't work due to missing MySQL support. Error in log file: `2025-02-04 08:34:06.106 sogod[637:100423] EOAdaptor: cannot find adaptor bundle: 'MySQL'`.

**Comment by iredmail** [comment]
@xbeaudouin @bozhinov 
Any update?

**Comment by bozhinov** [comment]
Sorry, I was not aware there was something pending on me. What is it ?

**Comment by iredmail** [comment]
- Mostly pending on @xbeaudouin :) We need help to improve 2 ports.
- hi @bozhinov could you help test other backends too?

**Comment by bozhinov** [comment]
Definitely. Let me know once ready for testing
I checked on the mysql option and it is incomplete. 
though if it is worth something in the current state I will test it.

I m not sure we are getting 6 more packages just cause of iredmail. 
I mean, the pkg build process is probably long enough as it is.

**Comment by iredmail** [comment]
> I checked on the mysql option and it is incomplete.

The incomplete part is the missed package flavors (sope, dovecot).
Can we get some help from other FreeBSD ports commitors?

**Comment by xbeaudouin** [comment]
So we need:
- sope@mysql
- sope@pgsql
- sope@ldapmysql
- dovecot@ldapmysql

Is that all we need?

**Comment by iredmail** [comment]
> So we need:
> 
> * sope@mysql
> * sope@pgsql
> * sope@ldapmysql
> * dovecot@ldapmysql
> 
> Is that all we need?

Exactly. :) @xbeaudouin

**Comment by xbeaudouin** [comment]
Sope : https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=284900 https://reviews.freebsd.org/D49056

**Comment by xbeaudouin** [comment]
For dovecot, I don't think we need a ldapmysql flavor since:
```
OPTIONS_DEFAULT=        GSSAPI_NONE LDAP LIBWRAP
```

And the flavor -mysql add just MYSQL to `OPTIONS_DEFAULT`:
```
.if ${FLAVOR:U} == mysql
OPTIONS_DEFAULT+=       MYSQL
.endif
```

**Comment by iredmail** [comment]
Hi @xbeaudouin 

Thanks for helping.

> For dovecot, I don't think we need a ldapmysql flavor since:

You made my day: https://github.com/iredmail/iRedMail/commit/660ebb926dce1a7d83f291334cc8c8a0fc549dda
Now all software except SOGo works in latest iRedMail git branch `pkg`.

About `devel/sope`, i found it supports ldap by default (same as Dovecot), so maybe we just need to build its `mysql` flavor to get both mysql and ldap support? for example `sope-mysql`. In this case your patch should be updated and drop `ldapmysql` flavor.

**Comment by xbeaudouin** [comment]
Well I have to add every flavors otherwise we can have an issue for sope. Waiting for maintainer feedback !

**Comment by iredmail** [comment]
Hi @xbeaudouin 

In case you didn't notice yet, there's a reply left in review: https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=284900
Could you help reply?

**Comment by iredmail** [comment]
Hi @xbeaudouin,

Do you have some time to check the review?

**Comment by xbeaudouin** [comment]
Hello,

acm@ from devel/sope etc... had pushed some new flavors : https://freshbsd.org/freebsd/ports/commit/f92a6c473dc6e6d0c778afd02d206b130325c52a

**Comment by iredmail** [comment]
> acm@ from devel/sope etc... had pushed some new flavors : https://freshbsd.org/freebsd/ports/commit/f92a6c473dc6e6d0c778afd02d206b130325c52a

I emailed him and requested this change. :)

**Comment by iredmail** [comment]
Hi @xbeaudouin @bozhinov 

- FreeBSD ports tree has been updated to build binary packages for sope / sogo with required flavors.
- iRedMail git branch `pkg` has been updated to use `Latest` pkg branch.

Could you please help test iRedMail installation?

**Comment by bozhinov** [comment]
Sure, what combinations of installation options do you want me to test?

**Comment by iredmail** [comment]
All backends, with all components enabled. :)

**Comment by iredmail** [comment]
Any update? @bozhinov

**Comment by bozhinov** [comment]
I m with the kids this weekend so it will have to take place the next one

**Comment by iredmail** [comment]
amd64 platform should work fine now, but no packages for arm64 / aarch64 yet.

**Comment by Meiko42** [comment]
First - thanks so much for the effort to use binaries for installation. Greatly appreciated!

I'm installing now on a FreeBSD 14.2-p3 system inside a VNET jail, Caddy reverse proxy in a separate VNET jail. Initial impression is that everything appears to be working, I'll spend some time tonight and tomorrow testing things out

**Comment by Meiko42** [comment]
The php_fpm process seems to stop every 2 days approximately, no noticeable pattern and no useful information in the logs :/

**Comment by iredmail** [comment]
> The php_fpm process seems to stop every 2 days approximately, no noticeable pattern and no useful information in the logs :/

- Do you mean php fpm process re-spawns or stopped completely?
- Is Roundcube webmail accessible (it relies on php-fpm)?

**Comment by Meiko42** [comment]
php_fpm stops completely, requires manual restart (`service php_fpm start`). Hasn't happened in the last couple days. 

When this happens Roundcube no longer loads, as expected

**Comment by Meiko42** [comment]
I should have set the php-fpm logging to debug earlier, that's done now. 

I'll leave that logging level enabled until this happens again, see if there's anything useful there. If there's any other troubleshooting tools I should implement for this let me know, otherwise I'll update when this issue pops up again.

**Comment by Meiko42** [comment]
php-fpm died again, no clear indication as to why from the debugging level logs. I checked my mail in Roundcube this morning and it was confirmed working, so this happened sometime in the last ~6hrs. I need to set up a script to notify when the process stops


```
root@mail:/ # ps -aux | grep -i php
root      96654  0.0  0.0   13836    2404  4  S+J  20:26     0:00.00 grep -i php
root@mail:/ # 
```


Entirety of `cat /var/log/php-fpm/php-fpm.log` 

```
root@mail:/ # cat /var/log/php-fpm/php-fpm.log
May 13 23:00:00 mail newsyslog[69406]: logfile turned over
root@mail:/ # 
```


Bottom of `zcat /var/log/php-fpm/php-fpm.log.0.gz`
```
127.0.0.1 -  13/May/2025:22:55:17 +0000 "POST /index.php" 200
May 13 22:55:43 mail syslogd: last message repeated 31 times
127.0.0.1 -  13/May/2025:22:56:11 +0000 "POST /index.php" 200
127.0.0.1 -  13/May/2025:22:56:17 +0000 "POST /index.php" 200
127.0.0.1 -  13/May/2025:22:57:12 +0000 "POST /index.php" 200
127.0.0.1 -  13/May/2025:22:57:17 +0000 "POST /index.php" 200
May 13 22:57:44 mail syslogd: last message repeated 118 times
May 13 22:58:12 mail syslogd: last message repeated 27 times
May 13 22:58:13 mail php-fpm[69747]: [DEBUG] fpm_pctl_perform_idle_server_maintenance(), line 398: [pool inet] currently 1 active children, 6 spare children, 7 running children. Spawning rate 1
127.0.0.1 -  13/May/2025:22:58:13 +0000 "POST /index.php" 200
May 13 22:58:14 mail php-fpm[69747]: [DEBUG] fpm_pctl_perform_idle_server_maintenance(), line 398: [pool inet] currently 0 active children, 7 spare children, 7 running children. Spawning rate 1
127.0.0.1 -  13/May/2025:22:58:17 +0000 "POST /index.php" 200
May 13 22:58:45 mail syslogd: last message repeated 30 times
127.0.0.1 -  13/May/2025:22:59:14 +0000 "POST /index.php" 200
127.0.0.1 -  13/May/2025:22:59:17 +0000 "POST /index.php" 200
May 13 23:00:00 mail newsyslog[69406]: logfile turned over
```

**Comment by iredmail** [comment]
No idea yet. :(

**Comment by Meiko42** [comment]
I've updated PHP to 8.3.21, will see if that resolves it.

**Comment by Meiko42** [comment]
Happy to report no sudden crashes of php_fpm since updating to 8.3.21. Will keep an eye on it through the rest of this week in case it's a fluke.

**Comment by Meiko42** [comment]
It's been a rough weekend - I briefly posted that 8.3.21 was good before remembering that's not actually the case, and deleted the message. 

What does appear to have resolved this is changing to PHP 8.4 latest (and associated Roundcube pkg). Continuing to monitor through the end of this week, no issues so far.

**Comment by iredmail** [comment]
Good news.

**Comment by Meiko42** [comment]
I spoke too soon, just crashed again. I'm at a loss on how to troubleshoot this. 

@bozhinov have you experienced issues with php_fpm stopping every 2-3 days when testing this? I'm curious if this is just a problem on my system.
